### DISCUSSION TITLE
Using Tectonic crates in rust project resulting in recompilation of of the entire dependency instead of just my project

### DISCUSSION THREAD

**Discussion by newholder** [tectonic-typesetting/tectonic#1091] [Q&A]
I am new to rust and i wanna try making some sort of API where user can upload a data and a template and it will generate beautiful looking pdf with that dynamic data, and i think latex is the way becaus of their beautiful result. So, i search arround and found tectonic. My goal is to embed it to my program. After wrangling with vcpkg. i get it to compile. But, everytime i change even one letter, and do cargo run, it has to compile the entire dependency. Why is that the case? How do i improve the compile time

**Comment by pkgw** [comment]
Just to be clear — you are editing the source code in your crate, which depends on the `tectonic` crate, and as a result stuff in the `tectonic` dependency graph needs to be recompiled? That definitely shouldn't be how things work ...

If you run `cargo build` twice in a row without editing *anything*, does it recompile anything?

What OS are you using? Are you using `cargo vcpkg` to set up Tectonic's dependencies?

The Rust build system will aggressivly recompile things if any environment variables related to the build change their values. So if you have an environment variable like `VCPKG_ROOT` set, you should make sure that your workflow always has it set to the same value, to minimize unneeded rebuilds.

**Reply by newholder** [reply]
I'm using windows with powershell. I didn't use cargo vcpkg to setup dependencies, just regular vcpkg. i run a script to setup the environment. 

set-env.ps1
```powershell
$env:TECTONIC_DEP_BACKEND = 'vcpkg'
$env:VCPKGRS_TRIPLET='x64-windows-static'
$env:RUSTFLAGS='-Ctarget-feature=+crt-static'
```

It does build correctly. But, if I run `cargo build` again, it will rebuild again if I change anything, and if the 1st `cargo build` and 2nd `cargo build` is run when changes aren't committed to git, without changes in between `cargo build`. it compiles the whole dependencies, not just tectonic like I'm running `cargo clean`. I also tried running it one by one in the terminal to see if maybe running it on a script messed it up somehow. Still the same result. But, if I do change > build > commit > build again, the 2nd build just finished instantly, as expected. 

I'll conduct experiments with various condition and update my reply as i found it

**Reply by pkgw** [reply]
If the behavior depends on whether you have committed changes to Git, this might not have to do with Tectonic specifically. I could imagine some kind of build script setup that will consider the environment to have been changed based the sort of output you get from `git status` ... but I'm not aware of anything within the Tectonic dependency graph that has that kind of behavior. Are you able to confirm that this behavior has something to do with Tectonic specifically?

**Reply by newholder** [reply]
I tried isolating any influence and try building it just using the built-in terminal, and editing the code using just Notepad, and magically, works as expected. It builds. It doesn't rebuild the entire tree when edited. I guess, this is either vscode or rust-analyzer messing it all up. I'd like for it to work on vscode. But I don't know how to accomplish it or any way to debug the cargo build process.

edit:
Poking around i noticed there's a verbose flag in the cargo build. Here's 3 consecutive run of cargo build with verbose

1st run: [output1.txt](https://github.com/tectonic-typesetting/tectonic/files/12711484/output1.txt)
2nd run: [output2.txt](https://github.com/tectonic-typesetting/tectonic/files/12711485/output2.txt)
3rd run: [output3.txt](https://github.com/tectonic-typesetting/tectonic/files/12711482/output3.txt)


i don't know how to decode this
